v8env/ts/fly/proxy.ts

```diff
@@ -0,0 +1,117 @@
+ */
+/**
+ * This generates a `fetch` like function for proxying requests to a given origin.
+ * When this function makes origin requests, it adds standard proxy headers like 
+ * `X-Forwarded-Host` and `X-Forwarded-For`. It also passes headers from the original
+ * request to the origin.
+ * @param origin A URL to an origin, can include a path to rebase requests.
+ * @param options Options and headers to control origin request.
+ */
+export default function proxy(origin: string | URL, options?: ProxyOptions) {
+  return function proxyFetch(req: RequestInfo, init?: RequestInit) {
+    if (!options) {
+      options = {}
+    }
+    const breq = buildProxyRequest(origin, options, req, init)
+    return fetch(breq)
+  }
+}
+
+interface FlyRequest extends Request {
+  url: string
+}
+
+/**
+ * @protected
+ * @hidden
+ * @param origin 
+ * @param options 
+ * @param req 
+ * @param init 
+ */
+export function buildProxyRequest(origin: string | URL, options: ProxyOptions, req: RequestInfo, init?: RequestInit) {
+
+  if (typeof req === "string") {
+    req = new Request(req)
+  }
+  const url = new URL(req.url)
+  let breq: FlyRequest | null = null
+
+  if (req instanceof Request) {
+    breq = req.clone()
+  } else {
+    breq = new Request(req)
+  }
+
+  if (typeof origin === "string") {
+    origin = new URL(origin)
+  }
+
+  url.hostname = origin.hostname
+  url.protocol = origin.protocol
+  url.port = origin.port
+
+  if (options.rewritePath && typeof options.rewritePath === 'string') {
+    // remove basePath so we can serve `onehosthame.com/dir/` from `origin.com/`
+    url.pathname = url.pathname.substring(options.rewritePath.length)
+    console.log("rewriting base path:", options.rewritePath, url.pathname)
+  }
+  if (origin.pathname && origin.pathname.length > 0) {
+    url.pathname = [origin.pathname.replace(/\/$/, ''), url.pathname.replace(/^\//, "")].join("/")
+  }
+  if (url.pathname.startsWith("//")) {
+    url.pathname = url.pathname.substring(1)
+  }
+
+  breq.url = url.toString()
+  // we extend req with remoteAddr
+  breq.headers.set("x-forwarded-for", (<any>req).remoteAddr)
+  breq.headers.set("x-forwarded-host", url.hostname)
+
+  if (options.headers) {
+    for (const h of Object.getOwnPropertyNames(options.headers)) {
+      const v = options.headers[h]
+      if (v === false) {
+        breq.headers.delete(h)
+      } else if (v && typeof v === "string") {
+        breq.headers.set(h, v)
+      }
+    }
+  }
+  return <Request>breq
+}
+
+/**
+ * Options for `proxy`.
+ */
+export interface ProxyOptions {
+  /**
+   * Replace this portion of URL path before making request to origin.
+   * 
+   * For example, this makes a request to `https://fly.io/path1/to/document.html`:
+   * ```javascript
+   * const opts = { rewritePath: "/path2/"}
+   * const origin = proxy("https://fly.io/path1/", opts)
+   * origin("https://somehostname.com/path2/to/document.html")
+   * ```
+   */
+  rewritePath?: string,
```

Comment:
I find this option confusing.

I usually associate "rewrite" like "replace". Needs 2 arguments in my opinion. Some kind of "from" and "to".

Also, a regex would be nice. Hell, I guess it should act just like `String.prototype.replace`!